[DPE-6650] Onboard mysql-rock

sinclert-canonical opened the pull request:
- [x] I have signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] I am following the [CONTRIBUTING](https://github.com/canonical/oci-factory/blob/main/CONTRIBUTING.md) guidelines

*Ping the @canonical/rocks team.*

---

## Description
This PR proposes the integration between the [mysql-rock](https://github.com/canonical/mysql-rock) repository (containing the image to be published to Dockerhub's Ubuntu namespace) and the OCI-Factory, in order to automatize such process.

My team and I are fairly new to the OCI-Factory shenanigans, so we are unsure what EOL se should use for images coming out of the `edge` channel. Looking at other repositories ([example 1](https://github.com/canonical/oci-factory/blob/main/oci/kafka/image.yaml), [example 2](https://github.com/canonical/oci-factory/blob/main/oci/zookeeper/image.yaml)), it seems having a past EOL date is perfectly fine so I proposed 2025-01-01 🤷🏻‍♂️ 

### Related issues
- Depends on issue https://github.com/canonical/oci-factory/issues/388.

cjdcordeiro commented on oci/mysql/documentation.yaml:
````diff
@@ -0,0 +1,80 @@
+    ```bash 
+    docker network create mysql-network
+    docker network connect mysql-network mysql-container
+    docker run -it --rm --network mysql-network ubuntu/mysql:8.0-24.04 mysql -h mysql-container -u root -p
````
this may need some work. as discussed in #388 , older images will be docker images, while newer versions will be rocks. the latter won't follow this deployment syntax. So please adjust the docs accordingly (see what was done in #387 )

sinclert-canonical replied:
I think I understood what you mean: there are no examples within the OCI folder of rockcraft built containers being run in interactive mode (`-it`). The container should run "as is", in a detached fashion, and if someone wants to connect to the mysql server within it, it must do so from outside the rock.

Is my understanding correct?

sinclert-canonical replied:
I think I understood what you meant: there are no examples within the OCI [folder](https://github.com/canonical/oci-factory/tree/main/oci) of rockcraft built containers being run in interactive mode (`-it`). The container should run "as is", in a detached fashion, and if someone wants to connect to the mysql server within it, it must do so from outside the rock.

Is my understanding correct?

cjdcordeiro replied:
not so much about the `-it` but rather that doing `docker run .... ubuntu/mysql:8.0-24.04 mysql` will return an error for this rock, since the entrypoint is Pebble and thus PID will result in `pebble enter mysql` (which is an error). Instead it should be `docker run .... ubuntu/mysql:8.0-24.04 --args <mysql-pebble-service> -h mysql-container  -u root -p`

sinclert-canonical replied:
I just had a conversation with @paulomach, and we are not entirely sure what you mean, as we could not find any onboarded image passing arguments to the underlying pebble service using the `--args`.

Do you want us to differentiate pre and post `8.0-24.04` usage in the documentation, similar to what folks did on the referenced [PR](https://github.com/canonical/oci-factory/pull/387) with OpenJDK 8/17 and OpenJDK 21? Please clarify.

sinclert-canonical replied:
Hey! Thanks for looking into this.

I discovered we were missing the entry-point command in the service (see [fixing PR](https://github.com/canonical/mysql-rock/pull/3) 🤦🏻‍♂️ ), as well as some config files that they were not correctly dumped.

As for the running command, I think it will be best to skip the explanation on how to connect from within the container, and focus on explaining how to get the image running. For this last case, just stating a `--run` after the image name would be enough to get the pebble services running. 

- For ubuntu/mysql < `8.0-24.04`:
    ```shell
    docker run --detach --name mysql-container \
        -p 30306:3306 \
        -e TZ=UTC  \
        -e MYSQL_ROOT_PASSWORD=My:S3cr3t \
        ubuntu/mysql:8.0-24.04
    ```

- For ubuntu/mysql > `8.0-24.04`:
    ```shell
    docker run --detach --name mysql-container \
        -p 30306:3306 \
        -e TZ=UTC  \
        -e MYSQL_ROOT_PASSWORD=My:S3cr3t \
        ubuntu/mysql:8.0-24.04 --run
    ```

cjdcordeiro replied:
Intead of `--run` you can just make `startup: enabled` for your mysql service.

Thanks for the fix. Let us know when the updated commit is added to this PR.

PS: on that PR, I noticed you are using a wrapper script as the pebble service command. I'd strongly suggest looking into improving that, cause such entrypoints are typically a bad Docker-induced practice, for Docker images who want to comprise multiple "modes" of running their container. Nginx is another good example of this, where you can run the container with the `nginx` command, or `nginx-debug` command, and there's a similar wrapper script to parse that CLI in and launch the corresponding processes. With Pebble, we should think differently, and instead of wrapper scripts, you should have multiple Pebble services, such that people can simply do `docker run mysql start <serviceOfChoice>`. Food for thought...

sinclert-canonical replied:
> Instead of --run you can just make startup: enabled for your mysql service.

Oh, then we do not even need to differentiate running commands, as the public MySQL rock was created leveraging pebble's `startup: enabled` syntax (see [code](https://github.com/canonical/mysql-rock/blob/3323392ef8a3a5e1606101ad4e4aed0af4807ce3/rockcraft.yaml#L16-L22)).

> I noticed you are using a wrapper script as the pebble service command. I'd strongly suggest looking into improving that, cause such entrypoints are typically a bad Docker-induced practice

We were already using an entry-point shell script for the MySQL Dockerhub images published < `8.0-24.04`. Again, the idea of the new image is to mimic, as much as possible, the way we were building the image with Docker, but with Rockcraft machinery.

If getting rid of the entry-point shell script is a hard requirement, then @paulomach should get involved.

cjdcordeiro replied:
well, I can't make that a hard requirement. it's your product in the end. but I'd definitely try to use Pebble as much as possible

paulomach replied:
Hi @cjdcordeiro , thanks for the inputs. We are sticking with the script for now.

cjdcordeiro replied:
ack

sinclert-canonical commented:
Just:
- Merged the MySQL-rock [PR](https://github.com/canonical/mysql-rock/pull/3) with the discussed fixes.
- Updated this PR to point to the MySQL-rock commit SHA containing them (see [these changes](https://github.com/canonical/oci-factory/pull/389/commits/93477ae0931269b7a3e922b4eebf9cfa726bbd60)).
- Updated this PR to remove the discintion between pre and post `24.04` tagged images (see [these changes](https://github.com/canonical/oci-factory/pull/389/commits/93477ae0931269b7a3e922b4eebf9cfa726bbd60)).

@cjdcordeiro let us know if you need anything else.

sinclert-canonical commented:
Just:
- Merged the MySQL-rock [PR](https://github.com/canonical/mysql-rock/pull/3) with the discussed fixes.
- Updated this PR to point to the MySQL-rock commit SHA containing them (see [these changes](https://github.com/canonical/oci-factory/pull/389/commits/93477ae0931269b7a3e922b4eebf9cfa726bbd60)).
- Updated this PR to remove the distinction between pre and post `24.04` tagged images (see [these changes](https://github.com/canonical/oci-factory/pull/389/commits/93477ae0931269b7a3e922b4eebf9cfa726bbd60)).

@cjdcordeiro let us know if you need anything else.

cjdcordeiro reviewed (approved):
Thanks for the changes. LGTM.

@zhijie-yang pls have a 2nd look and feel free to approve+merge

zhijie-yang commented on oci/mysql/image.yaml:
```diff
@@ -0,0 +1,11 @@
+    directory: .
+    release:
+      8.0-24.04:
+        end-of-life: "2025-01-01T00:00:00Z"
```
This EOL date is wrongly set to a past date. Please check your support timespan and update this image trigger. Otherwise, the downstream workflows won't get triggered to build and test this image.

sinclert-canonical replied:
👋🏻 Hey Zhijie! Thanks for que quick response.

From the PR description: 

> My team and I are fairly new to the OCI-Factory shenanigans, so we are unsure what EOL se should use for images coming out of the `edge` channel. Looking at other repositories ([example 1](https://github.com/canonical/oci-factory/blob/main/oci/kafka/image.yaml), [example 2](https://github.com/canonical/oci-factory/blob/main/oci/zookeeper/image.yaml)), it seems having a past EOL date is perfectly fine so I proposed 2025-01-01 🤷🏻‍♂️

Alternatively, I prepared a [PR](https://github.com/canonical/mysql-rock/pull/2) to automatize the updating of the MySQL rock OCI-Factory spec, when new releases of such image are triggered, setting a new EOL date each time. Would that be enough, or do you still want me to tweak the EOL date set in this PR?

zhijie-yang replied:
Clear. Sorry for overlooking this description. It's just from the view of the pipelines --- Setting the EOL to a past date makes most of the CI checks skipped, and we are missing the confirmation from the CI indicating if this rock is "good" to get merged. Since it won't get built with a trespassed EOL, I'll give my approval on this rock to have it onboarded (without building and publishing it). Checks for the production rocks will apply when you raise further PRs or trigger the workflows using the CLI client regarding this rock with future EOLs. Thanks for the rapid reply!

zhijie-yang reviewed (approved):
LGTM, thanks.
